src/agent/api.rs

```diff
@@ -0,0 +1,139 @@
+            token: token.to_string(),
+        }
+    }
+
+    fn build_request(&self, method: Method, url: &str) -> RequestBuilder {
+        let mut req = self.client
+            .request(method, &format!("{}/agent-api/{}", self.url, url));
+        req.header(header::Authorization(format!("token {}", self.token)));
+        req
+    }
+
+    fn retry<T, F: Fn(&Self) -> Result<T>>(&self, f: F) -> Result<T> {
+        loop {
+            match f(self) {
+                Ok(res) => return Ok(res),
+                Err(err) => {
+                    if let ErrorKind::ReqwestError(ref error) = *err.kind() {
+                        if error
+                            .get_ref()
+                            .map(|e| e.is::<::std::io::Error>())
+                            .unwrap_or(false)
+                        {
+                            warn!("connection to the server failed. retrying in a few seconds...");
+                            ::std::thread::sleep(::std::time::Duration::from_secs(RETRY_AFTER));
+                            continue;
+                        }
+                    }
+                    return Err(err);
+                }
+            }
+        }
+    }
+
+    pub fn config(&self) -> Result<AgentConfig> {
+        self.retry(|this| {
+            this.build_request(Method::Get, "config")
+                .send()?
+                .to_api_response()
+        })
+    }
+
+    pub fn next_experiment(&self) -> Result<Experiment> {
+        self.retry(|this| loop {
+            let resp: Option<_> = this.build_request(Method::Get, "next-experiment")
+                .send()?
+                .to_api_response()?;
+
+            if let Some(experiment) = resp {
+                return Ok(experiment);
+            }
+
+            ::std::thread::sleep(::std::time::Duration::from_secs(RETRY_AFTER));
+        })
+    }
+
+    pub fn record_result(
+        &self,
+        krate: &Crate,
+        toolchain: &Toolchain,
+        log: &[u8],
+        result: TestResult,
+        shas: &[(GitHubRepo, String)],
+    ) -> Result<()> {
+        self.retry(|this| {
+            let _: bool = this.build_request(Method::Post, "record-result")
+                .json(&json!({
+                    "crate": krate,
+                    "toolchain": toolchain,
+                    "result": result,
+                    "log": base64::encode(log),
+                    "shas": shas,
+                }))
+                .send()?
+                .to_api_response()?;
+            Ok(())
+        })
+    }
+
+    pub fn complete_experiment(&self) -> Result<()> {
+        self.retry(|this| {
+            let _: bool = this.build_request(Method::Post, "complete-experiment")
+                .send()?
+                .to_api_response()?;
+            Ok(())
+        })
+    }
+
+    pub fn heartbeat(&self) -> Result<()> {
+        self.retry(|this| {
+            let _: bool = this.build_request(Method::Post, "heartbeat")
```

Comment:
You've done such a great job of documenting the API it would be sad if this one didn't get any love :)

Reply:
Yeah, also `/record-result` was undocumented. Documented them and added some more information to the docs.